buildGoModule: building Go binaries in two phases

kalbasit opened the pull request:
###### Motivation for this change

This change provides a new builder called `buildGoModule` for building Go packages with [go modules](https://github.com/golang/go/wiki/Modules) support in mind. It does build through two phases:

1. The first phase consists of building the go-modules derivation: `go mod download` will be executed within the source directory to download all of the dependencies, and store them at `$GOPATH/pkg/mod`. After removing the impure directory `$GOPATH/pkg/mod/cache/vcs`, we persist `$GOPATH/pkg/mod` as the $out of this derivation.
2. The second phase consists of building the actual Go module, and this is done by setting `GOPROXY` to point to the download directory within `${go-modules}/cache/download` which instructs Go to skip any network call, and simply use the already-downloaded dependencies.

The benefit of such infrastructure is that we no longer need tools such as dep2nix or vgo2nix to parse the go.mod and go.sum in order to generate Nix dependency file.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date. [Rendered](https://kalbas.it/nixpkgs/pull/57080/doc/#sec-language-go).
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- Converted packages
	- [x] Move over `hugo`.
	- [x] Move over `pet`.
	- [x] Move over `jx`.
	- [x] Move over `gotools`.
    - [x] Move over `terminal-parrot`.
    - [x] Move over `todoist`.
    - [x] Move over `hetzner-kube`.
- Create new packages
	- [x] Create `mod`.

---

cc @ehmry @lethalman @fpletz @Mic92 @nlewo 

```
nix run -f https://github.com/kalbasit/nur-packages/archive/master.tar.gz nix-verify -c nix-verify -I nixpkgs=`pwd` --keep-going \
	-A jx \
	-A pet \
	-A hugo \
    -A gotools \
    -A terminal-parrot \
    -A todoist \
    -A hetzner-kube \
    -A mod
```

kalbasit committed:
buildGoPackage: move it under a different path

This change moves buildGoPackage from pkgs/development/go-modules to
pkgs/development/go-packages, so we can have buildGoModule at
pkgs/development/go-modules.

kalbasit committed:
jx: use buildGoModule instead of buildGoPackage

kalbasit committed:
buildGoModule: function for packaging Go modules

The function buildGoModule builds Go programs managed with Go modules. It builds
a Go module through a two phase build:

- An intermediate fetcher derivation. This derivation will be used to
  fetch all of the dependencies of the Go module.
- A final derivation will use the output of the intermediate derivation
  to build the binaries and produce the final output.

kalbasit committed:
pet: use buildGoModule instead of buildGoPackage

kalbasit commented:
I got `pet` to build this way, but there are still problems to solve:

- [ ] Go relies on `$GOPATH` to figure out where to store the cache, otherwise, it defaults to `$HOME/go`. It would be nice to have Go write to a different path instead.
- [ ] If `$GOPATH/go/pkg/mod/cache` does not exist, Go will attempt to reach to network even though all the dependencies are available under `$GOPATH/go/pkg/mod`. Leaving the `$GOPATH/go/pkg/mod/cache` will not work, at it contains VCS repositories that will change anytime upstream changes.

kalbasit committed:
hugo: use buildGoModule instead of buildGoPackage

kalbasit committed:
gotools: 2019-02-11 -> 2019-03-05

kalbasit committed:
gotools: use buildGoModule instead of buildGoPackage

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

Mic92 reviewed (changes requested):
I agree with the overall design. I just would keep the API of `buildGoModule` a bit more consistent with other builders and therefore easier to learn. So instead of introducing `fetchAttrs` and `buildAttrs` all meta data could be flattened into the derivation itself. For `buildRustPackage` we have an attribute called `cargoSha256`. For `buildGoModule` this could become `vendorSha256`.

Mic92 commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
> I agree with the overall design. I just would keep the API of `buildGoModule` a bit more consistent with other builders and therefore easier to learn. So instead of introducing `fetchAttrs` and `buildAttrs` all meta data could be flattened into the derivation itself. For `buildRustPackage` we have an attribute called `cargoSha256`. For `buildGoModule` this could become `vendorSha256`.

SGTM, thank you for the review! I'm working on flattening the attributes now.

I'm also aware that the build is failing, I was tired last night and had to push as-is so I can continue from my travel laptop this morning :)

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@GrahamcOfBorg build jx pet hugo gotools

kalbasit commented:
@Mic92 the PR is now ready for review! I've also updated the [rendered documentation](https://kalbas.it/nixpkgs/pull/57080/doc/#sec-language-go).

kalbasit committed:
terminal-parrot: indent and use version when fetching dependencies

kalbasit committed:
terminal-parrot: 1.1.0 -> 1.1.1 using buildGoModule

kalbasit committed:
todoist: use buildGoModule instead of buildGoPackage

kalbasit committed:
hetzner-kube: 0.3.2rc1 -> 0.4.1 using buildGoModule

kalbasit committed:
hetzner-kube: set the version via ldflags

Mic92 commented on pkgs/development/go-modules/generic/default.nix:
```diff
@@ -146,11 +135,9 @@ let
-          pushd "$NIX_BUILD_TOP/go/src" >/dev/null
-          find "$goPackagePath" -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort | uniq
-          popd >/dev/null
+          find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort | uniq
```
```suggestion
          find . -type f -name \*$type.go -exec dirname {} \; | grep -v "/vendor/" | sort -u
```

Mic92 commented on doc/languages-frameworks/go.xml:
```diff
@@ -3,12 +3,107 @@
+  <example xml:id='ex-buildGoModule'>
+   <title>buildGoModule</title>
+<programlisting>
+pet = buildGoModule rec {
```
By default buildGoPackage was a split output, $bin contains the executables
and $out contained the source. Since we did not had a use-case, source from $out was dropped.
Since then $out is empty for many packages. I wonder if for `buildGoModule` we just not start using multiple outputs, since we don't really need them.

kalbasit replied:
I've updated the function and fix gotools that got broken by this change.

kalbasit committed:
mod: init at 0.2.0

kalbasit commented:
@Mic92 I addressed your comments. I have also conducted some tests to verify with certainty that the modules downloaded are not affected by the changes upstream. This PR is now clear to be merged as well.

PTAL.
